backend/src/main/java/eu/clarin/switchboard/resources/DataResource.java

```diff
@@ -183,18 +194,46 @@ public Response getOutline(@Context HttpServletRequest request, @PathParam("id")
         if (fi == null) {
             return Response.status(Response.Status.NOT_FOUND).build();
         }
-        if (!fi.getProfile().toProfile().isMediaType("application/zip")) {
-            LOGGER.debug("getOutline returns: no content");
+        Map<String, Object> ret = new HashMap<>();
+        List<ZEntry> outline = null;
+
+        if (fi.getProfile().toProfile().isMediaType(Constants.MEDIATYPE_ZIP)) {
+            try (ZipFile zfile = new ZipFile(fi.getPath().toFile())) {
+                outline = zfile.stream()
+                        .filter(e -> !e.isDirectory() && e.getSize() > 0)
+                        .filter(e -> !e.getName().startsWith("__MACOSX/"))
+                        .map(e -> new ZEntry(e.getName(), e.getSize()))
+                        .collect(Collectors.toList());
+                if (outline.size() > MAX_ZIP_ENTRIES) {
+                    outline = outline.subList(0, (int) MAX_ZIP_ENTRIES);
+                    ret.put("outlineIsIncomplete", true);
+                }
+            } catch (ZipException xc) {
+                LOGGER.info("bad zip archive: " + xc.getMessage());
+                return Response.status(Response.Status.NOT_ACCEPTABLE).entity(xc.getMessage()).build();
+            }
+        } else if (fi.getProfile().toProfile().isMediaType(Constants.MEDIATYPE_TAR)) {
+            try (BufferedInputStream fis = new BufferedInputStream(new FileInputStream(fi.getPath().toFile()));
+                 TarArchiveInputStream tais = new TarArchiveInputStream(fis)) {
+                outline = new ArrayList<>();
+                for (TarArchiveEntry entry = tais.getNextTarEntry(); entry != null; entry = tais.getNextTarEntry()) {
+                    if (tais.canReadEntryData(entry) && !entry.isDirectory() && entry.getSize() > 0) {
+                        if (outline.size() >= MAX_ZIP_ENTRIES) {
+                            ret.put("outlineIsIncomplete", true);
+                            break;
+                        }
+                        outline.add(new ZEntry(entry.getName(), entry.getSize()));
+                    }
+                }
+            }
+        }
```

Comment:
It would take me a rather long time to figure out what is going on here.. maybe you can add some inline comments to make it easier to follow the logic?

Reply:
I refactored the code with additional 2 methods that make it more clear.

Reply:
Great, much more readable now